java/arcs/core/storage/StorageKeyParser.kt

```diff
@@ -18,31 +18,66 @@ package arcs.core.storage
  * If you modify the default set of storage keys in a test (using [addParser]), remember to call
  * [reset] in the tear-down method.
  */
-object StorageKeyParser {
-    private var parsers = DEFAULT_PARSERS.toMutableMap()
+interface StorageKeyManager {
+    fun parse(rawKeyString: String): StorageKey
+    fun addParser(parser: StorageKeyParser<*>)
+    fun reset(vararg initialSet: StorageKeyParser<*>)
+}
+
+/**
+ * The interface for a parser of a specific protocol type.
+ *
+ * By convention, when implementing [StorageKey], you should also include in the companion object
+ * a [PROTOCOL] field that defines the protocol for the name, and a [PARSER] that implements this
+ * interface.
+ *
+ * The companion object for this interface implements a thread-safe global instance of
+ * [StorageKeyManager], and that's currently what we use throughout the Arcs codebase for storage
+ * key management.
+ */
+interface StorageKeyParser<T : StorageKey> {
+    val protocol: String
+    fun parse(rawKeyString: String): T
+
+    /**
+     * Expose the [DefaultStorageKeyManager] via the [StorageKeyParser] type. We can later
+     * change the usage points to refer to [DefaultStorageKeyManager] directly, and remove this.
+     */
+    companion object : StorageKeyManager by DefaultStorageKeyManager
+}
+
+/** A global default thread-safe implementation of [StorageKeyManager]. */
+object DefaultStorageKeyManager : StorageKeyManager {
+    private val VALID_KEY_PATTERN = "^([\\w-]+)://(.*)$".toRegex()
+    private var parsers = mutableMapOf<String, StorageKeyParser<*>>()
 
     /** Parses a raw [key] into a [StorageKey]. */
-    fun parse(key: String): StorageKey {
-        val match = requireNotNull(VALID_KEY_PATTERN.matchEntire(key)) { "Illegal key: \"$key\"" }
+    override fun parse(rawKeyString: String): StorageKey {
+        val match =
+            requireNotNull(VALID_KEY_PATTERN.matchEntire(rawKeyString)) {
+                "Illegal key: \"$rawKeyString\""
+            }
 
         val protocol = match.groupValues[1]
         val contents = match.groupValues[2]
-        val parser =
-            requireNotNull(parsers[protocol]) { "Unknown protocol \"$protocol\" in \"$key\"" }
+        val parser = synchronized(this) {
+            requireNotNull(parsers[protocol]) {
+                "Unknown protocol \"$protocol\" in \"$rawKeyString\""
+            }
```

Comment:
This can wait until a later PR, but we should probably not include the key's full-string representation in the error message in case it contains PII in the future.

Reply:
I think this is happening in a few places, do you know if we have a task yet to resolve this?

Reply:
I don't think so.. I just thought of it as I was reviewing.

Reply:
I created b/165019603